dagster-io/dagster
# [auto run retry tags] 4/n -  update auto run reexecution to use `will_retry` and `auto_retry_run_id` tags

## Summary & Motivation
Updates the auto run reexecution daemon to look at the presence of a `will_retry` tag to determine if a run should be retried, rather than recomputing it from state. 

If a failed run does not have a `will_retry` tag, we recompute whether it should be retried to prevent against a deployment having different code and daemon versions

Associated internal pr:  https://github.com/dagster-io/internal/pull/12766

## How I Tested These Changes
new unit test for fallback behavior 

Manual testing:
Env: webserver/daemons running this branch, code server on an older version of dagster
```python
@dg.asset
def has_retry_policy(context):
    if context.run.parent_run_id is None:
        raise Exception("I failed")
    else:
        return 1 
```

Test 1
ran the asset - that run fails. Observe that the `will_retry` tag is not set. See the daemon pick up the run and launch a retry, indicating the fallback behavior worked. See that the `will_retry` tag has been added to the run, as has the `auto_retry_run_id` tag

Test 2
stop running the daemon. Run a few materializations of the asset, all fail. See that the `will_retry` tag is not added to any run. Start the daemon back up. The daemon processes the failed runs and retries them. See that the `will_retry` and `auto_retry_run_id` tags have been added to the original runs

Test 3 (also added a unit test for this case)
stop running the daemon. Materialize the asset, it fails. See that the `will_retry` tag is not added to the run. Stop the code server. Start the daemon back up. The idea here is to simulate the code server going down before the daemon processes the run. The daemon processes the run and sees it should retry, but cannot launch the retry because the code server is down. The daemon will not process this run again, so we need to make sure the `will_retry` tag is not True, since the run will never be retried. See that the `will_retry` tag on the run is `false`

## Changelog

The automatic run retry daemon has been updated so that there is a single source of truth for if a run will be retried and if the retry has been launched. Tags are now added to run at failure time indicating if the run will be retried by the automatic retry system. Once the automatic retry has been launched, the run ID of the retry is added to the original run.
